engine/src/main/java/org/pentaho/di/job/entries/ftp/JobEntryFTP.java

```diff
@@ -0,0 +1,1528 @@
+        if ( isDetailed() ) {
+          logDetailed( BaseMessages.getString( PKG, "JobEntryFTP.ChangedDir", realFtpDirectory ) );
+        }
+      }
+
+      // Create move to folder if necessary
+      if ( movefiles && !Utils.isEmpty( movetodirectory ) ) {
+        realMoveToFolder = environmentSubstitute( movetodirectory );
+        realMoveToFolder = normalizePath( realMoveToFolder );
+        // Folder exists?
+        boolean folderExist = true;
+        if ( isDetailed() ) {
+          logDetailed( BaseMessages.getString( PKG, "JobEntryFTP.CheckMoveToFolder", realMoveToFolder ) );
+        }
+        String originalLocation = ftpclient.pwd();
+        try {
+          // does not work for folders, see PDI-2567: folderExist=ftpclient.exists(realMoveToFolder);
+          // try switching to the 'move to' folder.
+          ftpclient.chdir( realMoveToFolder );
+          // Switch back to the previous location.
+          if ( isDetailed() ) {
+            logDetailed( BaseMessages.getString( PKG, "JobEntryFTP.CheckMoveToFolderSwitchBack", originalLocation ) );
+          }
+          ftpclient.chdir( originalLocation );
+        } catch ( Exception e ) {
+          folderExist = false;
+          // Assume folder does not exist !!
+        }
+
+        if ( !folderExist ) {
+          if ( createmovefolder ) {
+            ftpclient.mkdir( realMoveToFolder );
+            if ( isDetailed() ) {
+              logDetailed( BaseMessages.getString( PKG, "JobEntryFTP.MoveToFolderCreated", realMoveToFolder ) );
+            }
+          } else {
+            logError( BaseMessages.getString( PKG, "JobEntryFTP.MoveToFolderNotExist" ) );
+            exitjobentry = true;
+            NrErrors++;
+          }
+        }
+      }
+
+      if ( !exitjobentry ) {
+        // Get all the files in the current directory...
+        FTPFile[] ftpFiles = ftpclient.dirDetails( null );
+
+        if ( isDetailed() ) {
+          logDetailed( BaseMessages.getString( PKG, "JobEntryFTP.FoundNFiles", String.valueOf( ftpFiles.length ) ) );
+        }
+
+        // set transfertype ...
+        if ( binaryMode ) {
+          ftpclient.setType( FTPTransferType.BINARY );
+          if ( isDetailed() ) {
+            logDetailed( BaseMessages.getString( PKG, "JobEntryFTP.SetBinary" ) );
+          }
+        } else {
+          ftpclient.setType( FTPTransferType.ASCII );
+          if ( isDetailed() ) {
+            logDetailed( BaseMessages.getString( PKG, "JobEntryFTP.SetAscii" ) );
+          }
+        }
+
+        // Some FTP servers return a message saying no files found as a string in the filenlist
+        // e.g. Solaris 8
+        // CHECK THIS !!!
+
+        if ( ftpFiles.length == 1 ) {
+          String translatedWildcard = environmentSubstitute( wildcard );
+          if ( !Utils.isEmpty( translatedWildcard ) ) {
+            if ( ftpFiles[0].getName().startsWith( translatedWildcard ) ) {
+              throw new FTPException( ftpFiles[0].getName() );
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Exception [FTPException] is explicitly thrown on this line but is caught in the outer catch clause with generic exception on line [1153].  Consider adding a catch clause of this specific exception in the outer try statement.